sweepbatcher/presigned.go

```diff
@@ -0,0 +1,457 @@
+				"for feeRate %v: %w", fr, err)
+		}
+
+		// If the feerate is high enough, set locktime to prevent
+		// broadcasting such a transaction too early by mistake.
+		if fr >= highFeeRate {
+			tx.LockTime = highFeeRateLockTime
+		}
+
+		// Try to presign this transaction.
+		err = presigner.Presign(ctx, tx, batchAmt)
+		if err != nil {
+			return fmt.Errorf("failed to presign unsigned tx %v "+
+				"for feeRate %v: %w", tx.TxHash(), fr, err)
+		}
+
+		// If fee was clamped, stop here, because fee rate won't grow.
+		if fee < feeForWeight {
+			break
+		}
+	}
+
+	return nil
+}
+
+// publishPresigned creates sweep transaction using a custom transaction signer
+// and publishes it. It returns the fee of the transaction, and an error (if
+// signing and/or publishing failed) and a boolean flag indicating signing
+// success. This mode is incompatible with an external address.
+func (b *batch) publishPresigned(ctx context.Context) (btcutil.Amount, error,
+	bool) {
+
+	// Sanity check, there should be at least 1 sweep in this batch.
+	if len(b.sweeps) == 0 {
+		return 0, fmt.Errorf("no sweeps in batch"), false
+	}
+
+	// Make sure that no external address is used.
+	for _, sweep := range b.sweeps {
+		if sweep.isExternalAddr {
+			return 0, fmt.Errorf("external address was used with " +
+				"a custom transaction signer"), false
+		}
+	}
+
+	// Cache current height and desired feerate of the batch.
+	currentHeight := b.currentHeight
+	feeRate := b.rbfCache.FeeRate
+
+	// Append this sweep to an array of sweeps. This is needed to keep the
+	// order of sweeps stored, as iterating the sweeps map does not
+	// guarantee same order.
+	sweeps := make([]sweep, 0, len(b.sweeps))
+	for _, sweep := range b.sweeps {
+		sweeps = append(sweeps, sweep)
+	}
+
+	// Cache the destination address.
+	address, err := b.getSweepsDestAddr(ctx, sweeps)
+	if err != nil {
+		return 0, fmt.Errorf("failed to find destination address: %w",
+			err), false
+	}
+
+	// Construct unsigned batch transaction.
+	tx, weight, _, fee, err := constructUnsignedTx(
+		sweeps, address, currentHeight, feeRate,
+	)
+	if err != nil {
+		return 0, fmt.Errorf("failed to construct tx: %w", err),
+			false
+	}
+
+	// Adjust feeRate, because it may have been clamped.
+	feeRate = chainfee.NewSatPerKWeight(fee, weight)
+
+	// Calculate total input amount.
+	batchAmt := btcutil.Amount(0)
+	for _, sweep := range sweeps {
+		batchAmt += sweep.value
+	}
+
+	// Determine the current minimum relay fee based on our chain backend.
+	minRelayFee, err := b.wallet.MinRelayFee(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get minRelayFee: %w", err),
+			false
+	}
+
+	// Get a signed transaction. It may be either new transaction or a
+	// pre-signed one.
+	const presignedOnly = false
+	signedTx, err := b.cfg.presignedHelper.SignTx(
+		ctx, tx, batchAmt, minRelayFee, feeRate, presignedOnly,
+	)
+	if err != nil {
+		return 0, fmt.Errorf("failed to sign tx: %w", err),
```

Comment:
is it trying to sign when `presignedOnly` is passed in?

Reply:
`presignedOnly = false` in this call, so it tries to sign interactively and if it fails, goes to the store of presigned transactions.